File: src/main/java/org/mifos/fineract/services/DocumentsApi.java
package org.mifos.fineract.services;

import okhttp3.MultipartBody;
import org.mifos.fineract.models.DeleteEntityTypeEntityIdDocumentsResponse;
import org.mifos.fineract.models.GetEntityTypeEntityIdDocumentsResponse;
import org.mifos.fineract.models.PostEntityTypeEntityIdDocumentsResponse;
import org.mifos.fineract.models.PutEntityTypeEntityIdDocumentsResponse;
import retrofit2.http.*;
import rx.Observable;

import java.util.List;

public interface DocumentsApi {
    /**
     * Create a Document
     * Note: A document is created using a Multi-part form upload   Body Parts  name :  Name or summary of the document  description :  Description of the document  file :  The file to be uploaded  Mandatory Fields :  file and description
     *
     * @param entityType    entityType (required)
     * @param entityId      entityId (required)
     * @param contentLength Content-Length (optional)
     * @param file          file (optional)
     * @param name          name (optional)
     * @param description   description (optional)
     * @return Observable&lt;PostEntityTypeEntityIdDocumentsResponse&gt;
     */
    @retrofit2.http.Multipart
    @POST("{entityType}/{entityId}/documents")
    Observable<PostEntityTypeEntityIdDocumentsResponse> createDocument(
            @retrofit2.http.Path("entityType") String entityType, @retrofit2.http.Path("entityId") Long entityId, @retrofit2.http.Header("Content-Length") Long contentLength, @retrofit2.http.Part MultipartBody.Part file, @retrofit2.http.Part("name") String name, @retrofit2.http.Part("description") String description
    );

    /**
     * Remove a Document
     *
     * @param entityType entityType (required)
     * @param entityId   entityId (required)
     * @param documentId documentId (required)
     * @return Observable&lt;DeleteEntityTypeEntityIdDocumentsResponse&gt;
     */
    @Headers({
            "Content-Type:application/json"
    })
    @DELETE("{entityType}/{entityId}/documents/{documentId}")
    Observable<DeleteEntityTypeEntityIdDocumentsResponse> deleteDocument(
            @retrofit2.http.Path("entityType") String entityType, @retrofit2.http.Path("entityId") Long entityId, @retrofit2.http.Path("documentId") Long documentId
    );

    /**
     * Retrieve Binary File associated with Document
     * Request used to download the file associated with the document  Example Requests:  clients/1/documents/1/attachment   loans/1/documents/1/attachment
     *
     * @param entityType entityType (required)
     * @param entityId   entityId (required)
     * @param documentId documentId (required)
     * @return Observable&lt;Void&gt;
     */
    @Headers({
            "Content-Type:application/json"
    })
    @GET("{entityType}/{entityId}/documents/{documentId}/attachment")
    Observable<Void> downloadFile(
            @retrofit2.http.Path("entityType") String entityType, @retrofit2.http.Path("entityId") Long entityId, @retrofit2.http.Path("documentId") Long documentId
    );

    /**
     * Retrieve a Document
     * Example Requests:  clients/1/documents/1   loans/1/documents/1   client_identifiers/1/documents/1?fields&#x3D;name,description
     *
     * @param entityType entityType (required)
     * @param entityId   entityId (required)
     * @param documentId documentId (required)
     * @return Observable&lt;GetEntityTypeEntityIdDocumentsResponse&gt;
     */
    @Headers({
            "Content-Type:application/json"
    })
    @GET("{entityType}/{entityId}/documents/{documentId}")
    Observable<GetEntityTypeEntityIdDocumentsResponse> getDocument(
            @retrofit2.http.Path("entityType") String entityType, @retrofit2.http.Path("entityId") Long entityId, @retrofit2.http.Path("documentId") Long documentId
    );

    /**
     * List documents
     * Example Requests:  clients/1/documents  client_identifiers/1/documents  loans/1/documents?fields&#x3D;name,description
     *
     * @param entityType entityType (required)
     * @param entityId   entityId (required)
     * @return Observable&lt;List&lt;GetEntityTypeEntityIdDocumentsResponse&gt;&gt;
     */
    @Headers({
            "Content-Type:application/json"
    })
    @GET("{entityType}/{entityId}/documents")
    Observable<List<GetEntityTypeEntityIdDocumentsResponse>> retreiveAllDocuments(
            @retrofit2.http.Path("entityType") String entityType, @retrofit2.http.Path("entityId") Long entityId
    );

    /**
     * Update a Document
     * Note: A document is updated using a Multi-part form upload  Body Parts name Name or summary of the document description Description of the document file The file to be uploaded
     *
     * @param entityType    entityType (required)
     * @param entityId      entityId (required)
     * @param documentId    documentId (required)
     * @param contentLength Content-Length (optional)
     * @param file          file (optional)
     * @param name          name (optional)
     * @param description   description (optional)
     * @return Observable&lt;PutEntityTypeEntityIdDocumentsResponse&gt;
     */
    @retrofit2.http.Multipart
    @PUT("{entityType}/{entityId}/documents/{documentId}")
    Observable<PutEntityTypeEntityIdDocumentsResponse> updateDocument(
            @retrofit2.http.Path("entityType") String entityType, @retrofit2.http.Path("entityId") Long entityId, @retrofit2.http.Path("documentId") Long documentId, @retrofit2.http.Header("Content-Length") Long contentLength, @retrofit2.http.Part MultipartBody.Part file, @retrofit2.http.Part("name") String name, @retrofit2.http.Part("description") String description
    );

}

File: src/main/kotlin/org/openapitools/client/apis/DocumentsApi.kt
package org.openapitools.client.apis

import okhttp3.MultipartBody
import org.openapitools.client.models.DeleteEntityTypeEntityIdDocumentsResponse
import org.openapitools.client.models.GetEntityTypeEntityIdDocumentsResponse
import org.openapitools.client.models.PostEntityTypeEntityIdDocumentsResponse
import org.openapitools.client.models.PutEntityTypeEntityIdDocumentsResponse
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Path

interface DocumentsApi {
    /**
     * Create a Document
     * Note: A document is created using a Multi-part form upload   Body Parts  name :  Name or summary of the document  description :  Description of the document  file :  The file to be uploaded  Mandatory Fields :  file and description
     * Responses:
     *  - 200: Not Shown (multi-part form data)
     *
     * @param entityType entityType
     * @param entityId entityId
     * @param contentLength Content-Length (optional)
     * @param dateFormat  (optional)
     * @param description  (optional)
     * @param locale  (optional)
     * @param name  (optional)
     * @param uploadedInputStream  (optional)
     * @return [PostEntityTypeEntityIdDocumentsResponse]
     */
    @Multipart
    @POST("v1/{entityType}/{entityId}/documents")
    suspend fun createDocument(
        @Path("entityType") entityType: String,
        @Path("entityId") entityId: Long,
        @Header("Content-Length") contentLength: Long? = null,
        @Part("dateFormat") dateFormat: String? = null,
        @Part("description") description: String? = null,
        @Part("locale") locale: String? = null,
        @Part("name") name: String? = null,
        @Part uploadedInputStream: MultipartBody.Part? = null
    ): PostEntityTypeEntityIdDocumentsResponse

    /**
     * Remove a Document
     *
     * Responses:
     *  - 200: OK
     *
     * @param entityType entityType
     * @param entityId entityId
     * @param documentId documentId
     * @return [DeleteEntityTypeEntityIdDocumentsResponse]
     */
    @DELETE("v1/{entityType}/{entityId}/documents/{documentId}")
    suspend fun deleteDocument(
        @Path("entityType") entityType: String,
        @Path("entityId") entityId: Long,
        @Path("documentId") documentId: Long
    ): DeleteEntityTypeEntityIdDocumentsResponse

    /**
     * Retrieve Binary File associated with Document
     * Request used to download the file associated with the document  Example Requests:  clients/1/documents/1/attachment   loans/1/documents/1/attachment
     * Responses:
     *  - 200: Not Shown: The corresponding Binary file
     *
     * @param entityType entityType
     * @param entityId entityId
     * @param documentId documentId
     * @return [Unit]
     */
    @GET("v1/{entityType}/{entityId}/documents/{documentId}/attachment")
    suspend fun downloadFile(
        @Path("entityType") entityType: String,
        @Path("entityId") entityId: Long,
        @Path("documentId") documentId: Long
    ): Unit

    /**
     * Retrieve a Document
     * Example Requests:  clients/1/documents/1   loans/1/documents/1   client_identifiers/1/documents/1?fields&#x3D;name,description
     * Responses:
     *  - 200: OK
     *
     * @param entityType entityType
     * @param entityId entityId
     * @param documentId documentId
     * @return [GetEntityTypeEntityIdDocumentsResponse]
     */
    @GET("v1/{entityType}/{entityId}/documents/{documentId}")
    suspend fun getDocument(
        @Path("entityType") entityType: String,
        @Path("entityId") entityId: Long,
        @Path("documentId") documentId: Long
    ): GetEntityTypeEntityIdDocumentsResponse

    /**
     * List documents
     * Example Requests:  clients/1/documents  client_identifiers/1/documents  loans/1/documents?fields&#x3D;name,description
     * Responses:
     *  - 200: OK
     *
     * @param entityType entityType
     * @param entityId entityId
     * @return [kotlin.collections.List<GetEntityTypeEntityIdDocumentsResponse]
     */
    @GET("v1/{entityType}/{entityId}/documents")
    suspend fun retrieveAllDocuments(
        @Path("entityType") entityType: String,
        @Path("entityId") entityId: Long
    ): List<GetEntityTypeEntityIdDocumentsResponse>

    /**
     * Update a Document
     * Note: A document is updated using a Multi-part form upload  Body Parts name Name or summary of the document description Description of the document file The file to be uploaded
     * Responses:
     *  - 200: Not Shown (multi-part form data)
     *
     * @param entityType entityType
     * @param entityId entityId
     * @param documentId documentId
     * @param contentLength Content-Length (optional)
     * @param dateFormat  (optional)
     * @param description  (optional)
     * @param locale  (optional)
     * @param name  (optional)
     * @param uploadedInputStream  (optional)
     * @return [PutEntityTypeEntityIdDocumentsResponse]
     */
    @Multipart
    @PUT("v1/{entityType}/{entityId}/documents/{documentId}")
    suspend fun updateDocument(
        @Path("entityType") entityType: String,
        @Path("entityId") entityId: Long,
        @Path("documentId") documentId: Long,
        @Header("Content-Length") contentLength: Long? = null,
        @Part("dateFormat") dateFormat: String? = null,
        @Part("description") description: String? = null,
        @Part("locale") locale: String? = null,
        @Part("name") name: String? = null,
        @Part uploadedInputStream: MultipartBody.Part? = null
    ): PutEntityTypeEntityIdDocumentsResponse

}
